In [10]:
import init_notebooks
from data_loader import load_data
from train_model import Train_Model

X_train, X_test, y_train, y_test = load_data()

In [11]:
train = Train_Model()

In [12]:
models = train.models()

Prépare tous les modèles de régression que tu veux tester sur ton dataset.

Paramètres choisis pour équilibre entre performance et overfitting.

Permet d’avoir un pipeline clair pour entraîner, évaluer et comparer plusieurs modèles rapidement.

In [13]:
train.select_model(models, X_train, y_train, X_test, y_test)

                           MSE        R2
LinearRegression  3.397926e+07  0.781130
Ridge             3.401915e+07  0.780873
Lasso             3.397956e+07  0.781128
RandomForest      2.203743e+07  0.858051
SVR               1.662463e+08 -0.070838
GradientBoosting  1.968366e+07  0.873212
XGBoost           1.894900e+07  0.877944


Évalue tous les modèles définis précédemment sur le même dataset.

Utilise un pipeline pour appliquer le préprocessing automatiquement.

Calcule MSE et R² pour comparer les performances. 

Choix du meilleur modèle : XGBoost , meilleur compromis MSE faible et R² élevé.

In [14]:
import joblib
from sklearn.pipeline import Pipeline
from xgboost import XGBRegressor

preprocessor = joblib.load("../preprocessor/preprocessor.pkl")

pipe = Pipeline(steps=[
    ("preprocess", preprocessor),       
     ("model", XGBRegressor(
        n_estimators=200,
        learning_rate=0.1,
        max_depth=3,
        random_state=42
    ))
])


pipe.fit(X_train, y_train)


joblib.dump(pipe, "../models/best_model.pkl")

['../models/best_model.pkl']

Prépare un pipeline prêt à la production : préprocessing + modèle.

Entraîne le meilleur modèle identifié (XGBoost).

Sauvegarde le pipeline pour prédictions futures sans refaire le preprocessing.

In [15]:
loaded_model = joblib.load("../models/best_model.pkl")

loaded_model.score(X_test, y_test) 


0.87794429063797

le modèle explique environ 87 % de la variance des charges médicales